### PR TITLE
Dynamic Charts for PID Tuning Page

### PR DESCRIPTION
**Pull request by GaryKeeble**
Guy's

I have coded a change which shows the expo charts using the betaflight calculations ported over to javascript.

I have taken it as far as I can at the moment; but changing the rates and expo settings only changes the chart after you have saved them and gone off and back onto the PID tuning tab.

I have designed it to allow for the rcData from the MSP to move a ball on the chart as you move the radio so you can see the curve in operation but I don't know how to setup and get the MSP commands to get the values (so the ball is fixed at 1500 - midpoint at the moment).

Chart shows the curve in deg/s.

Not sure you would want to merge this one, but it might help as a start point....

### CODE REVIEW

**Comment by borisbstyle** [comment]
It has conflicts now, but sounds good!

**Comment by borisbstyle** [comment]
If you solve the conflicts I will merge it and release

**Comment by nathantsoi** [comment]
nice work @garykeeble, i can take a look at making this update in realtime.
can we also merge
https://github.com/betaflight/betaflight-configurator/pull/57 for the next
release @borisbstyle?

File: tabs/pid_tuning.js [hunk 1]
```diff
@@ -498,3 +531,171 @@ TABS.pid_tuning.cleanup = function (callback) {
         callback();
     }
 };
+
+/* Wrap whole function in an independant class */
+
+function ExpoChart(canvas, rcData, rcExpo, rcRate, deadband, midrc, axisRate, superExpoActive) {
```
Comment (mikeller) [inline tabs/pid_tuning.js]: You should move these parameters (except for the canvas) into the `refresh` function. This way you can update the curve without recreating the object.

File: tabs/pid_tuning.js [hunk 2]
```diff
@@ -399,18 +399,51 @@ TABS.pid_tuning.initialize = function (callback) {
...
-                    rcYawCurveElement = $('.yaw_curve canvas').get(0);
-
-		drawRateCurve(rateElement, expoElement, rcCurveElement);
-		drawRateCurve(rateElement, yawExpoElement, rcYawCurveElement);
+				pitchRollCurve.refresh(1500 /*rcData*/, 
+									   parseFloat(expoElement.val()) * 100/*rcExpo*/, 
+									   parseFloat(rateElement.val()) * 100/*rcRate*/, 
+									   0/*deadband*/, 
+									   1500/*midrc*/, 
+									   70/*axisRate*/,
```
Comment (mikeller) [inline tabs/pid_tuning.js]: These should be set from actual values.

**Comment by mikeller** [comment]
Hi @GaryKeeble.

Great work, looks good. I made a couple of remarks how this could be made to update every time the input changes (you don't need to hook it up to MSP, as the data is already available in the DOM). Maybe we'd want to put the axis labels into html though, instead of recalculating / redrawing them every time the input changes.

Let me know if you want to do the changes, otherwise I am happy to take it from here and hook it up into the update handler.

**Comment by antstorm** [comment]
@mikeller seems like PID tab is getting a lot of attention :) Just pushed my PR #59 with a 3D model for rates preview.

What do you guys think about extracting Rates into a separate tab?

**Comment by mikeller** [comment]
@antstorm: See my comment in #59.

**Comment by borisbstyle** [comment]
guys the current master has broken configuration tab

@mikeller 
Can you rebase this one based on 3D model?

**Comment by antstorm** [comment]
@borisbstyle what's exactly broken?

**Comment by mikeller** [comment]
@borisbstyle: Will look into this tonight. Will also look into making it update on input changes.

**Comment by borisbstyle** [comment]
configuration tab not loading

**Comment by antstorm** [comment]
Odd, just pulled `master` locally, loads fine here. Any errors in dev tools?

**Comment by borisbstyle** [comment]
Weirdd It simply doesnt load on mine. 1.5 does work properly

**Comment by borisbstyle** [comment]
wait maybe I messed up something locally

**Comment by antstorm** [comment]
Any local changes?

**Comment by borisbstyle** [comment]
yeah looks like I had something messed up.

**Comment by borisbstyle** [comment]
@GaryKeeble 
Can you rebase this one?

**Comment by GaryKeeble** [comment]
@borisbstyle 
I'll re-base the code and update; no problem.

@mikeller 
I always wanted the chart to be dynamic but was unsure as to how to link to the dynamic information. Thanks for the suggestions; I'll start by just getting it fixed for merging but feel free to get it right for the update handler etc. I have only just started looking at the configurator so someone who is more familiar with the coding style and functions would obviously be better at this stage.

**Comment by mikeller** [comment]
@GaryKeeble: Good approach, I'll wait with changes then until you've fixed it and it's merged. I've got some more ideas, like combining multiple graphs into one, and I'll do them alongside with the updating.

**Comment by borisbstyle** [comment]
Also it would be good to hide the charts for older releases.

Also different chart for situation with super expo rates and without would
be interresting?
Op 30 jun. 2016 08:11 schreef "Michael Keller" <notifications@github.com>:

**Comment by mikeller** [comment]
@borisbstyle: Sorry, forgot to mention that, but to show either the new or the old graph based on firmware version was part of my plan to clean up the elements on the PID tab.

Re having graphs for super expo on / off, good idea. I would think giving users a toggle to toggle between rates with and without super expo is probably best, this way they can easily use the graph to help them decide if they want to use them or not. (There should also be a help text explaining where they can activate them.)

@GaryKeeble: That's one more thing, instead of deleting the old values can you just comment them out, so I can later on re-enable them. Or better, you could make them switchable with a

```
if (CONFIG.flightControllerIdentifier == "BTFL" && semver.gte(CONFIG.flightControllerVersion, "2.9.0")) {
```

**Comment by skaman82** [comment]
@borisbstyle The PID Tab ist getting more cluttered and ugly every day :(  Why do you keep adding things that are not styled properly/look unfinished or are even not working yet properly? 

Could anyone please include this table (see picture) into the show/hide unused pid action so it will only show if other PIDs are visible?
bildschirmfoto 2016-06-30 um 10 27 06

**Comment by borisbstyle** [comment]
@skaman82 Well its not going to be released till its finished and till it already looks good. I do want to have profile tuning options as much as possible in the same tab. Its the graphs making it cluttered.

Some things have to go or at least get smaller. Like for example the rate tool should probably just be a pop up. To calculate rates you would have to press the button to activate pop up with all rate graphs and moving model.

Other option is to make the graphs much smaller than now. Especially the 3D model can be really small if you ask me

**Comment by skaman82** [comment]
I just cleaned it up a bit. But there is still much work to do.
bildschirmfoto 2016-06-30 um 11 43 19

**Comment by skaman82** [comment]
there is also something broken: 
bildschirmfoto 2016-06-30 um 11 54 02

**Comment by mikeller** [comment]
@skaman82: Will look into it. Is this in master?

**Comment by skaman82** [comment]
@mikeller yes

**Comment by mikeller** [comment]
@borisbstyle: As a suggestion: How about you keep adding stuff until everything you want is in there, then you stop merging pull requests and let us know, and we can clean up and consolidate the UI?

**Comment by mikeller** [comment]
@skaman82: fixed in #63

Will look into combining the pitch / roll / yaw curves into one graph now.

**Comment by borisbstyle** [comment]
@mikeller 
Yes putting those in one single graph would be the best thing, with small 3D model next to it!

This is really the first startup for betaflight-configurator and obviously the moment where we need to search for the best workflow.
But thats actually what I am doing now.....just features features features and than cleanup.

The pending list:
- Rate calculator
- 3D model
- Add Vbat_pid_compensation checkbox somewhere

**Comment by GaryKeeble** [comment]
Guy's

This pull request is moving much too fast for me to be able to keep up (with the day job and all). I have tried to rebase the pull request and merge but my inadequate knowledge of Git is letting me down; Is there any chance I can hand over the modification to you guys to complete to your satisfaction?

I know that @nathantsoi has made some significant progress with mapping real-world values into my initial submission and clearly the rest of you have it well in hand.

Of course, I'll help as best I can...

File: _locales/en/messages.json [hunk 3]
```diff
@@ -526,7 +529,13 @@
     "featureTRANSPONDERTip": {
         "message": "Configure via the Race Transponder tab after enabling."
     },
-    
+    "featureSUPEREXPO_RATES": {
+        "message": "Rate value adds instead of rate also Super Expo. Mid stick stays same. Rc rate is always liniear"
```
Comment (mikeller) [inline _locales/en/messages.json]: Not wanting to be nitpicking, but it's `linear`.

File: _locales/en/messages.json [hunk 4]
```diff
@@ -93,6 +93,9 @@
     "tabTransponder": {
         "message": "Race Transponder"
     },
+    "tabOsd": {
+        "message": "OSD"
+    },
```
Comment (mikeller) [inline _locales/en/messages.json]: Is this related to your PR, or is this just an artefact of making it mergeable?

**Comment by mikeller** [comment]
@GaryKeeble: There are a lot of changes in your PR now that are not related to the new rates graph in the PID tab, so I suspect something went wrong in the rebase. If it's all right with you, I will cherry - pick the work you did for the new graph, and put it in a form that works to display either the new graph (for new firmware) or the old graph.

**Comment by nathantsoi** [comment]
that would be my 1/2 done commit that got merge into @GaryKeeble's fork @mikeller. see: https://github.com/GaryKeeble/betaflight-configurator/pull/2

it would be awesome if you want to continue on this @mikeller

**Comment by mikeller** [comment]
@nathantsoi: Yeah, sure. I wanted to do some reformatting of the legacy curves anyway, and I'll see if I can get the new curves to display in the same canvas, and display curves for pitch / roll / yaw on one canvas anyway.

Btw just tried @borisbstyle's default PID settings on my quad today, after my old settings got the wobbles after an update to 3.0, and they're awesome. Flies rock steady just on default PIDs. :-)

**Comment by nathantsoi** [comment]
awesome on the curves. and yes! i've been flying 3.0 for a while and it's outstanding! @borisbstyle... amazing :)

https://www.youtube.com/watch?v=qyPDDBMAk8o

even since the first releases of betaflight, i couldn't get over how amazing my quads fly on stock pids. once i fly it and feel how amazing it is, i dont think it can't get any better, but then, somehow boris makes it better.

**Comment by GaryKeeble** [comment]
@nathantsoi @mikeller 
Guy's

Really sorry about this; I merged @nathantsoi's code this morning so that I could take a look today if I got a chance at work!

**Comment by mikeller** [comment]
@GaryKeeble: No reason to be sorry really, concurrent versioning is a really challenging concept, often times more challenging than good software engineering.

I see this every time at my day job, and I've come to the conclusion that there is only two kinds of software engineers: Those who've stuffed up concurrent versioning, and those who haven't been in software engineering for long enough.

**Comment by borisbstyle** [comment]
Yeah we all do this between our normal job :D. No need for sorry.

**Comment by mikeller** [comment]
@GaryKeeble: I have grafted the code for the rates curve into #72. Thank you so much for your work, it was very useful, it just turned out to be easier to graft than to fix up this pull request, as there was a bunch of other things I wanted to do around the rate curves.

**Pull request closed without merging**